lang/wasm_instance_cache.hh

```diff
@@ -113,6 +130,39 @@ public:
 
     void remove(const db::functions::function_name& name, const std::vector<data_type>& arg_types) noexcept;
 
+    // Wasmtime instances hold references to modules, so the module can only be dropped
+    // when all instances are dropped. For a given module, we can have at most one
+    // instance for each scheduling group.
+    // This function is called each time a new instance is created for a given module.
+    // If there were no instances for the module before, i.e. this module was not
+    // compiled, the module is compiled and the size of the compiled code is added
+    // to the total size of compiled code. If the total size of compiled code exceeds
+    // the maximum size as a result of this, the function will evict modules until
+    // there is enough space for the new module. If it is not possible, the function
+    // will throw an exception. If this function succeeds, the counter of instances
+    // for the module is increased by one.
+    void track_module_ref(wasmtime::Module& module, wasmtime::Engine& engine);
+
+    // This function is called each time an instance for a given module is dropped.
+    // If the counter of instances for the module reaches zero, the module is dropped
+    // and the size of the compiled code is subtracted from the total size of compiled code.
+    void remove_module_ref(wasmtime::Module& module) noexcept;
+
+    // When a WASM UDF is executed, a separate stack is first allocated for it.
+    // This stack is used by the WASM code and it is not tracked by the seastar allocator.
+    // This function will evict cached modules until the stack can be allocated. If enough
+    // memory can't be freed, the function will throw an exception.
+    void reserve_wasm_stack();
+
+    // This function should be called after a WASM UDF finishes execution. Its stack is then
+    // destroyed and this function accounts for the freed memory.
+    void free_wasm_stack() noexcept;
```

Comment:
Nit: Is there a reason why those methods are public and not private? They seem to be internal to how the instance cached works. The methods `track_module_ref`/`remove_module_ref` now should probably be exclusively used by `module_handle`, so they could also be made private and `module_handle` could become a friend of `instance_cache`.